src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlRetry/SqlRetryService.cs

```diff
@@ -0,0 +1,147 @@
+            };
+
+        private readonly IsExceptionRetriable _defaultIsExceptionRetriable = DefaultIsExceptionRetriable;
+        private readonly bool _defaultIsExceptionRetriableOff;
+        private readonly IsExceptionRetriable _customIsExceptionRetriable;
+        private readonly int _maxRetries;
+        private readonly int _retryMillisecondsDelay;
+
+        public SqlRetryService(
+            IOptions<SqlRetryServiceOptions> sqlRetryServiceOptions,
+            SqlRetryServiceDelegateOptions sqlRetryServiceDelegateOptions)
+        {
+            EnsureArg.IsNotNull(sqlRetryServiceOptions?.Value, nameof(sqlRetryServiceOptions));
+            if (sqlRetryServiceOptions.Value.RemoveTransientErrors != null)
+            {
+                _transientErrors.ExceptWith(sqlRetryServiceOptions.Value.RemoveTransientErrors);
+            }
+
+            if (sqlRetryServiceOptions.Value.AddTransientErrors != null)
+            {
+                _transientErrors.UnionWith(sqlRetryServiceOptions.Value.AddTransientErrors);
+            }
+
+            _maxRetries = sqlRetryServiceOptions.Value.MaxRetries;
+            _retryMillisecondsDelay = sqlRetryServiceOptions.Value.RetryMillisecondsDelay;
+
+            EnsureArg.IsNotNull(sqlRetryServiceDelegateOptions, nameof(sqlRetryServiceDelegateOptions));
+            _defaultIsExceptionRetriableOff = sqlRetryServiceDelegateOptions.DefaultIsExceptionRetriableOff;
+            _customIsExceptionRetriable = sqlRetryServiceDelegateOptions.CustomIsExceptionRetriable;
+        }
+
+        public delegate bool IsExceptionRetriable(Exception ex);
+
+        public delegate Task RetriableAction();
+
+        private static bool DefaultIsExceptionRetriable(Exception ex)
+        {
+            if (ex is SqlException sqlEx)
+            {
+                if (sqlEx.Number == 121 && sqlEx.Message.Contains("an error occurred during the pre-login handshake", StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+
+                if (sqlEx.ToString().Contains("login failed", StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private bool RetryTest(Exception ex)
+        {
+            if (ex is SqlException sqlEx && _transientErrors.Contains(sqlEx.Number))
+            {
+                return true;
+            }
+
+            if (!_defaultIsExceptionRetriableOff && _defaultIsExceptionRetriable(ex))
+            {
+                return true;
+            }
+
+            if (_customIsExceptionRetriable?.Invoke(ex) == true)
+            {
+                return true;
+            }
+
+            if (ex.IsRetriable())
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        public async Task ExecuteWithRetries(RetriableAction action)
```

Comment:
>public async Task ExecuteWithRetries(RetriableAction action) [](http://example.com/codeflow?start=8&length=60)

Please add back reader method so people do not make common mistakes with IEnumerable and need to call next result.

Reply:
Done